src/libs/actions/IOU.ts

```diff
@@ -7035,9 +7035,19 @@ function canApproveIOU(iouReport: OnyxTypes.OnyxInputOrEntry<OnyxTypes.Report>,
     const iouSettled = ReportUtils.isSettled(iouReport?.reportID);
     const reportNameValuePairs = ReportUtils.getReportNameValuePairs(iouReport?.reportID);
     const isArchivedReport = ReportUtils.isArchivedRoom(iouReport, reportNameValuePairs);
-    const unheldTotalIsZero = iouReport && iouReport.unheldTotal === 0;
+    let isTransactionBeingScanned = false;
+    const reportTransactions = TransactionUtils.getAllReportTransactions(iouReport?.reportID);
+    for (const transaction of reportTransactions) {
```

Comment:
Ah there's a problem here, this still will allow approvals of 0-value scans (e.g. if you upload a junk receipt where the amount can't be read).

https://github.com/user-attachments/assets/10ea8cb7-91d5-4c19-90fe-36ed0c7ed473

Reply:
@jjcoffee That's true and expected given my solution. As stated in the other issue which implemented the `unheldTotalIsZero` logic, the issue's (https://github.com/Expensify/App/issues/48590) the expected result states:

> Approver should not be able to approve scanning expense

None of the issues state that we shouldn't allow `Approve` when the transaction amount is 0, but that we shouldn't allow `Approve` while a scan transaction is currently scanning.

To confirm this I asked the C+ / CME in the other issue https://github.com/Expensify/App/issues/48590#issuecomment-2430420547 but didn't get a response so I assumed that the above expected result is the way to go.

**Why should we allow `Approve` when amount is 0 ?**
Because when a scan transaction with blurry image fails and amount is 0, you get an error on the field that you should add an amount, which allows you to add an amount if current is 0 then `Approve` - otherwise you deliberately `Approve` with 0 amount, on purpose.

##

If you have more context that I'm missing here which confirms that we shouldn't show the `Approve` button if the transaction amount is 0, please let me know and I'll add that to the PRs logic - otherwise, if there are no other issues with the logic I think we're good to move forward with this.

Reply:
> otherwise you deliberately Approve with 0 amount, on purpose.

@ikevin127 I can't think of any reason why you'd want to approve an expense with a 0 amount. There's nothing explicitly wrong with it but it just feels a bit messy that it's even possible, I think!

I'm also assuming we don't want to allow this both from the title of https://github.com/Expensify/App/issues/48590 as well as the PR's [test step](https://github.com/Expensify/App/pull/50016):

> verify that approver can't approve the scan expense as 0.00 amount

Reply:
@jjcoffee The problem and pitfall in that PRs logic is that `canApproveIOU` is used in `MoneyReportHeader` and `ReportPreview` to show the `Approve` button for multiple expenses, because of this we can't hide `Approve` only for a transaction with amount 0 while showing it for the other transaction which is approvable (has all required fields completed).

That's why the `unheldTotalIsZero` logic added by the other PR was hiding the `Approve` button even for transactions what were not of type scan with 0 amount (our issue).

Say we implement this logic:

```ts
let isAmountMissing = false;
    let isTransactionBeingScanned = false;
    const reportTransactions = TransactionUtils.getAllReportTransactions(iouReport?.reportID);
    for (const transaction of reportTransactions) {
        const hasReceipt = TransactionUtils.hasReceipt(transaction);
        const isReceiptBeingScanned = TransactionUtils.isReceiptBeingScanned(transaction);

        // If transaction has receipt (scan) and its receipt is being scanned, we shouldn't be able to Approve
        if (hasReceipt && isReceiptBeingScanned) {
            isTransactionBeingScanned = true;
        }

        // Change isAmountMissing to true only if transaction amount is 0 (eg. failed scan)
        if (TransactionUtils.isAmountMissing(transaction)) {
            isAmountMissing = true;
        }
    }
    return isCurrentUserManager && !isOpenExpenseReport && !isApproved && !iouSettled && !isArchivedReport && !isTransactionBeingScanned && !isAmountMissing;
```
And we have a report with 2 transactions, 1 is manual and has amount and 1 is a failed scan (amount 0), with the logic above we won't see the `Approve` button at all.

<img width="500" alt="Screenshot 2024-10-29 at 15 14 54" src="https://github.com/user-attachments/assets/1cb288bb-8313-4b55-b05a-43989e01a10f">

Whereas with the current logic of this PR, say the manual transaction is on hold and we also have 1 failed scan transaction (amount 0), we will see the `Approve` button on `ReportPreview` and `MoneyReportHeader`, which when clicked will ask the user if they want to approve amount 0 (only the failed scan transaction) or both (failed scan and the held manual $7), which nobody would do, meaning to approve only the $0, in a real-world scenario right ?

<img width="500" alt="Screenshot 2024-10-29 at 15 19 26" src="https://github.com/user-attachments/assets/c0566989-bd77-4cdf-9265-73083956e8da">

To change this logic to your expectations would require rethinking how `canApproveIOU` works in `MoneyReportHeader` and `ReportPreview`, instead of showing `Approve` on a group of expenses (ReportPreview / MoneyReportHeader) we would have to implement logic to only show the button per individual transaction, which is definitely out of scope for our issue.

Please let me know what do you think we should do with this!

Reply:
Thanks for your thoughts @ikevin127! 

> we have a report with 2 transactions, 1 is manual and has amount and 1 is a failed scan (amount 0), with the logic above we won't see the Approve button at all.

Agree that we wouldn't want this, in this case and if the manual transaction is held I guess we'd ideally want to hide the `Approve only $0.00` button.

@marcochavezf Bringing you in here to chime in on whether we're okay to keep showing the `Approve` button for a failed scan (e.g. blurry image) with a **$0.00** amount. 

The main issue is that if you have another expense that's held, after tapping the `Approve` button, the first option shows as `Approve only $0.00`. That feels like dodgy UX to me, but I agree with @ikevin127 that fixing it is probably out of scope for this issue, since we wouldn't want to disable approving entirely as that would block approving the entire report (the second option in the screenshot below) and this would require a bit of a rethink of how to display that popup.

![image](https://github.com/user-attachments/assets/a3f92d07-5eca-4f41-81e7-3a0f38fa622f)

Just to note that the issue that introduced the regression had this as a test step:

> verify that approver can't approve the scan expense as 0.00 amount

So wanted to be sure we can go ahead despite that.

Reply:
## Breakdown - My two 🪙 

If you ask me, the changes of this PR are basically fixing our issue plus the expected result of this issue:
- https://github.com/Expensify/App/issues/48590

the problem is that the PR for the issue above:
- https://github.com/Expensify/App/pull/50817

went out of scope, because the issue stated clearly in its **expected result**:

> Approver should not be able to approve scanning expense

instead, the contributor tried to do more than the **expected result**, this is stated in PRs testing steps:

> verify that approver can't approve the scan expense as 0.00 amount

instead of simply fulfilling the issue's **expected result**. This led to our issue:
- https://github.com/Expensify/App/issues/50932

I'm assuming this happened because of lack of understanding of the codebase / logic and root cause.

##

**To conclude**: with our current logic it is not possible to only show the `Approve` button for individual transactions as I explained in https://github.com/Expensify/App/pull/51503#discussion_r1821606033, so this PR is both fixing its issue and the other issue where the **expected result** was to not be able to `Approve` a while an expense is still scanning.

From my proposal, this is how this would look:

| Can't approve while scanning |
| ------------- |
| <video src="https://github.com/user-attachments/assets/2da13c1c-1a73-45c3-a1aa-958313aea332"> |

##

If we want to rethink the logic and make the `Approve` button show per individual transaction, I think that deserves a separate task as it's way out of scope for any of the issues mentioned above.

cc @marcochavezf

Reply:
Thanks for bringing this up, guys! I agree that approving a report with an amount of 0 after scanning is not a good UX. Also, understandably it can be considered out-of-scope because it's an edge case we couldn't foresee. 

Given that we'd need to wait until this PR is deployed to staging/production to report it as a bug, what would be the solution to fix it here? We can adjust the price of the original issue by justifying the out-of-scope edge case.

Reply:
@marcochavezf I appreciate the input 🙏 

> I agree that approving a report with an amount of 0 after scanning is not a good UX. Also, understandably it can be considered out-of-scope because it's an edge case we couldn't foresee.

1. Approving a report with an amount of 0 after scanning was allowed, present and possible on prod / staging even before PRs https://github.com/Expensify/App/pull/50817 changes and wasn't reported as an issue. Therefore the fact that the current PR fixes our issue while maintaining PRs https://github.com/Expensify/App/pull/50817 changes its a good thing which I could've left out as our issue's scope wasnt to improve or correct the other PRs logic but I did because otherwise that issue would've been reintroduced.
2. To reinforce point (1.), it is for sure out of scope since up until me bringing up the fact that this PRs fix handles both our issue and maintaining the other PRs logic, this wasn't even an issue (never reported as one).

**Given all this, here are the 2 roads I'm willing to move forward with right now regarding this PR:**
1. Merge this PR as is since it fixes both our issue while maintaining the expected reult fix for issue https://github.com/Expensify/App/issues/48590.
2. Drop everything, close the PR and leave it to somebody else to handle this + any out of scope work.

**Reasoning**
I'm not willing to go out of scope here regardless of extra compensation simply because the issue of `approving a report with an amount of 0 after scanning` was present on prod / staging even before PR https://github.com/Expensify/App/pull/50817, therefore it should not be a problem or reported as a regression of this PR - IMO it should be a `New Feature` issue since it requires lots of changes throughout the app.

Reply:
> Given that we'd need to wait until this PR is deployed to staging/production to report it as a bug, what would be the solution to fix it here?

Given that it's a bit of an edge case, and it was also previously possible (to approve a $0 amount), I think we can just proceed with this PR and report the bug after.

Reply:
Got it, thanks for the additional context @ikevin127! Yeah, if approving a report with an amount of 0 after scanning was already allowed then it makes sense to report it as a bug and continue with this PR as-is